File: src/main/java/com/baidu/algorithm/datastructure/_208_Implement_Trie.py
class TrieNode:
    def __init__(self):
        self.is_word = False
        self.child = collections.defaultdict(TrieNode)

class Trie(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()


    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        cur = self.root
        for letter in word:
            cur = cur.child[letter]

        cur.is_word = True

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        cur = self.root
        for letter in word:
            cur = cur.child.get(letter)
            if cur is None:
                return False

        return cur.is_word

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        cur = self.root
        for letter in prefix:
            cur = cur.child.get(letter)
            if cur is None:
                return False

        return True

File: src/main/java/com/baidu/algorithm/datastructure/_208_Implement_Trie.java
/**
 * Copyright (C) 2016 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.datastructure;

/**
 * _208_Implement_Trie
 *
 * @author xuhaoran01
 */
public class _208_Implement_Trie {

    public static void main(String[] args) {

        String word = "hello";
        String prefix = "hel";

        Trie obj = new Trie();
        obj.insert(word);
        boolean param_2 = obj.search(word);
        boolean param_3 = obj.startsWith(prefix);
    }
}

class Trie {

    private TrieNode root;

    /**
     * Initialize your data structure here.
     */
    public Trie() {
        root = new TrieNode();
    }

    /**
     * Inserts a word into the trie.
     */
    public void insert(String word) {

        TrieNode cur = root;

        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (cur.child[c - 'a'] == null) {
                cur.child[c - 'a'] = new TrieNode();
            }

            cur = cur.child[c - 'a'];
        }

        cur.isWord = true;
    }

    /**
     * Returns if the word is in the trie.
     */
    public boolean search(String word) {

        TrieNode cur = root;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (cur.child[c - 'a'] == null) {
                return false;
            }

            cur = cur.child[c - 'a'];
        }

        return cur.isWord;
    }

    /**
     * Returns if there is any word in the trie that starts with the given prefix.
     */
    public boolean startsWith(String prefix) {

        TrieNode cur = root;
        for (int i = 0; i < prefix.length(); i++) {
            char c = prefix.charAt(i);
            if (cur.child[c - 'a'] == null) {
                return false;
            }

            cur = cur.child[c - 'a'];
        }

        return true;
    }
}

class TrieNode {
    public boolean isWord;
    public TrieNode[] child;

    public TrieNode() {
        this.child = new TrieNode[26];
    }
}